File: tests/test_reflect.py
import pytest

try:
    import test_reflect_ext as t
    def needs_reflect(x):
        return x
except ImportError:
    needs_reflect = pytest.mark.skip(reason="C++26 reflection support required")


@needs_reflect
def test01_enum():
    assert t.Enum.A.name == 'A'
    assert t.Enum.B.name == 'B'
    assert t.Enum.C.name == 'C'
    assert t.Enum.A.value == 0
    assert t.Enum.B.value == 1
    assert t.Enum.C.value == 2


@needs_reflect
def test02_default_ctor():
    s = t.Struct()
    assert s.i == 0
    assert s.d == 0.0
    assert s.s == ''


@needs_reflect
def test03_param_ctor():
    s = t.Struct(3, 1.5)
    assert s.i == 3
    assert s.d == 1.5


@needs_reflect
def test04_data_member_rw():
    s = t.Struct()
    s.i = 7
    s.d = 2.5
    s.s = 'hello'
    assert s.i == 7
    assert s.d == 2.5
    assert s.s == 'hello'


@needs_reflect
def test05_const_method():
    s = t.Struct(3, 1.5)
    assert s.get_i() == 3
    assert abs(s.sum() - 4.5) < 1e-10


@needs_reflect
def test06_nonconst_method():
    s = t.Struct()
    s.set_i(10)
    assert s.get_i() == 10


@needs_reflect
def test07_method_overload():
    s = t.Struct(0, 1.0)
    assert abs(s.overloaded(2.0) - 3.0) < 1e-10
    assert abs(s.overloaded(2, 3) - 6.0) < 1e-10


@needs_reflect
def test08_static_const():
    assert t.Struct.static_const == 42


@needs_reflect
def test09_static_mutable():
    before = t.Struct.static_mut
    t.Struct()
    assert t.Struct.static_mut == before + 1


@needs_reflect
def test10_static_method():
    assert t.Struct.create_count() >= 0


@needs_reflect
def test11_ns_class():
    a = t.A()
    a.x = 1
    a.y = 2.0
    assert a.x == 1
    assert a.y == 2.0

    b = t.B()
    b.s = 'test'
    assert b.s == 'test'


@needs_reflect
def test12_ns_enum():
    assert t.E.X.name == 'X'
    assert t.E.Z.name == 'Z'
    assert t.E.X.value == 0
    assert t.E.Z.value == 2


@needs_reflect
def test13_free_fn():
    assert abs(t.free_fn(1.0, 2.0) - 3.0) < 1e-10


@needs_reflect
def test14_free_fn_overload():
    assert abs(t.free_fn(1.0, 2.0, 3.0) - 6.0) < 1e-10


@needs_reflect
def test15_string_member():
    n = t.Nested()
    n.name = 'abc'
    assert n.name == 'abc'


@needs_reflect
def test16_vector_member():
    n = t.Nested()
    n.items = [1, 2, 3]
    assert n.items == [1, 2, 3]


@needs_reflect
def test17_nested_type():
    n = t.Nested()
    s = t.Struct(5, 1.0)
    n.inner = s
    assert n.inner.i == 5
    assert n.inner.d == 1.0


@needs_reflect
def test18_private_members_hidden():
    m = t.Mixed()
    m.pub_i = 7
    m.pub_s = 'hi'
    assert m.pub_i == 7
    assert m.pub_s == 'hi'
    assert m.get_pub() == 7

    assert not hasattr(m, 'priv_i')
    assert not hasattr(m, 'priv_d')
    assert not hasattr(m, 'priv_method')
    assert not hasattr(m, 'prot_i')

File: tests/test_reflect.cpp
#include <nanobind/nb_reflect.h>
#include <nanobind/stl/vector.h>
#include <nanobind/stl/string.h>
#include <cmath>
#include <string>
#include <vector>

namespace nb = nanobind;

namespace reflect_test {

enum class Enum { A, B, C };

struct Struct {
    int i;
    double d;
    std::string s;

    static constexpr int static_const = 42;
    static int static_mut;

    Struct() : i(0), d(0), s() { ++static_mut; }
    Struct(int i, double d) : i(i), d(d), s() { ++static_mut; }

    int get_i() const { return i; }
    void set_i(int v) { i = v; }
    double sum() const { return i + d; }

    double overloaded(double x) const { return d + x; }
    double overloaded(int x, int y) const { return d + x + y; }

    static int create_count() { return static_mut; }
};

int Struct::static_mut = 0;

struct Nested {
    std::string name;
    int value;
    std::vector<int> items;
    Struct inner;
};

class Mixed {
public:
    int pub_i;
    std::string pub_s;

    Mixed() : pub_i(0), pub_s(), priv_i(0), priv_d(0) {}

    int get_pub() const { return pub_i; }

private:
    int priv_i;
    double priv_d;

    void priv_method() {}

protected:
    int prot_i;
};

namespace ns {

struct A {
    int x;
    double y;
};

struct B {
    std::string s;
};

enum class E { X, Y, Z };

double free_fn(double a, double b) { return a + b; }
double free_fn(double a, double b, double c) { return a + b + c; }

} // namespace ns

} // namespace reflect_test

NB_MODULE(test_reflect_ext, m) {
    nb::reflect_<^^reflect_test>(m);
}
